### PR TITLE
[bndtools explorer] Refresh workspace from disk

### LINKED ISSUE #4413: Syncing the Eclipse Workspace with the disk workspace
I want to make a sync button in the Bndtools explorer that deletes projects not in the bnd workspace and adds any that are there.

@gamerson Can you give me a pointer how to import the bnd project as an Eclipse project? And any special things how to delete a project?

### LINKED ISSUE #4434: Workspace with only errors that warn that the project is not build
I've seen several situations where a whole workspace is ok, but the error from bnd that it has not been build is still there.

Do we need that error?

### PR DESCRIPTION
**Pull request by pkriens**
When the workspace is refreshed from the Bndtools 
explorer, it will do the following:

- Stop autobuild
- Add any projects that are in the bnd workspace
  but not in the Eclipse workspace
- Delete any projects in the Eclipse workspace
  that are located in the bnd workspace but are no
  longer found
- Clean all projects the bnd way 
- Update _all_ the .classpath files
- Refresh the whole workspace
- Set the workspace build order to the bnd 
  build workspace build order
- Set concurrent build jobs to 10
- Do a clean build
- Restore autobuild

This fixes #4413

This patch also fixes  #4434. When compiler errors
are detected, the error is marked directly in the
Eclipse workspace. This used to be reported as an
error on the Project. There was a bug in generate
that did not clear the bnd errors and then this
caused the problem that some fixed errors were still
counted.

Signed-off-by: Peter Kriens <Peter.Kriens@aqute.biz>

### CODE REVIEW

**Comment by pkriens** [comment]
Open issues I have is that I completely replace the .classpath file with a correct one:

- Is there a use case to have non-bnd info in the .classpath?
- Maybe this should be on a higher level that always automatically creates the .classpath?
-

**Comment by kriegfrj** [comment]
Im sorry, i might not get to review this until next week.

**Review by kriegfrj — CHANGES_REQUESTED** [review]
> Open issues I have is that I completely replace the .classpath file with a correct one:
> 
> * Is there a use case to have non-bnd info in the .classpath?

In the past, I have run m2e and bndtools side-by-side in the same project so that I could use the maven cxf-codegen-plugin. I never quite got it working properly. But I think it would be a good idea for us to leave this option open. Our users might want to get creative!

> * Maybe this should be on a higher level that always automatically creates the .classpath?

I'm leaning towards this (I raised exactly that in my review). Using Eclipse APIs to manipulate .classpath (rather than manipulating it directly) future-proofs the implementation, and also means that Eclipse is aware of the changes and can update its in-memory models, trigger updates on any watchers, etc.

File: bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java [hunk 1]
```diff
@@ -271,29 +270,30 @@
 					if (model.isNoBundles()) {
 						buildLog.basic("-nobundles was set, so no build");
 						buildLog.setFiles(0);
-						return report(model, markers);
+						return dependsOn;
```
[the remark this answers is not in the document]
Comment (kriegfrj) [reply bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java:273]: Actually, you could also change the last call to `report()` (L320) and get rid of `report()` too.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/eclipse/EclipseUtil.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import java.io.File;
+import java.net.URI;
+
+import aQute.bnd.build.Project;
+import aQute.bnd.osgi.Constants;
+import aQute.lib.io.IO;
+import aQute.lib.tag.Tag;
+
+public class EclipseUtil {
```
[the remark this answers is not in the document]
Comment (gamerson) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/eclipse/EclipseUtil.java:11]: I agree that writing this file by hand may run into compatibility problems with other plugins or future versions of eclipse.

[the remark this answers is not in the document]
Comment (gamerson) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/eclipse/EclipseUtil.java:11]: Agree, i think using Eclipse API for writing .classpath would be preferable.

[the remark this answers is not in the document]
Comment (pkriens) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/eclipse/EclipseUtil.java:11]: Well, I'll take my chances. We are using a small subset and this file format has been stable for 20 years. We do regular updates so we find any issues quickly.

**Pull request merged**